File: rust/crates/gribjump/benches/gribjump_bench.rs
//! Benchmarks for the gribjump crate.
//!
//! Run with: `cargo bench --package gribjump`
//!
//! These benchmarks measure FFI overhead, not gribjump core performance.
//! Benchmark names use `rust_` prefix for comparison with C++ benchmarks.

use criterion::{Criterion, black_box, criterion_group, criterion_main};
use gribjump::{ExtractionRequest, GribJump, Range};

/// Benchmark `GribJump` handle creation.
fn bench_handle_creation(c: &mut Criterion) {
    c.bench_function("rust_handle_creation", |b| {
        b.iter(|| black_box(GribJump::new().expect("failed to create handle")));
    });
}

/// Benchmark `ExtractionRequest` creation.
fn bench_request_creation(c: &mut Criterion) {
    let ranges = vec![
        Range::new(0, 11).expect("valid range"),
        Range::new(20, 31).expect("valid range"),
    ];

    c.bench_function("rust_request_creation", |b| {
        b.iter(|| {
            black_box(ExtractionRequest::new(
                "class=rd,expver=xxxx,stream=oper,date=20230508,time=1200",
                ranges.clone(),
                "gridhash",
            ))
        });
    });
}

/// Benchmark Range creation.
fn bench_range_creation(c: &mut Criterion) {
    c.bench_function("rust_range_creation", |b| {
        b.iter(|| black_box(Range::new(0, 101).expect("valid range")));
    });
}

/// Benchmark request creation with many ranges.
fn bench_request_many_ranges(c: &mut Criterion) {
    let ranges: Vec<Range> = (0..100)
        .map(|i| Range::new(i * 10, i * 10 + 5).expect("valid range"))
        .collect();

    c.bench_function("rust_request_100_ranges", |b| {
        b.iter(|| {
            black_box(ExtractionRequest::new(
                "class=rd,expver=xxxx,stream=oper,date=20230508,time=1200",
                ranges.clone(),
                "gridhash",
            ))
        });
    });
}

criterion_group!(
    benches,
    bench_handle_creation,
    bench_request_creation,
    bench_range_creation,
    bench_request_many_ranges,
);

criterion_main!(benches);

File: rust/crates/gribjump/benches/cpp/gribjump_bench.cpp
// C++ benchmarks for gribjump using Google Benchmark
//
// Uses the native gribjump::GribJump C++ API (same as what CXX wraps).
//
// Build: python3 scripts/bench_compare.py --build
// Run: python3 scripts/bench_compare.py --cpp
//
// Benchmark names use `cpp/` prefix for comparison with Rust benchmarks.
// These benchmarks measure FFI overhead, not gribjump core performance.

#include <benchmark/benchmark.h>
#include <eckit/runtime/Main.h>
#include <gribjump/GribJump.h>
#include <gribjump/LibGribJump.h>
#include <vector>

// Initialize eckit Main singleton (required for gribjump)
// This is done via static initialization before benchmarks run
namespace {
struct EckitInitializer {
    EckitInitializer() {
        // Use a minimal argc/argv for eckit initialization
        static const char* argv[] = {"gribjump_bench", nullptr};
        eckit::Main::initialise(1, const_cast<char**>(argv));
    }
};
static EckitInitializer eckit_init;
}  // namespace

static void BM_HandleCreation(benchmark::State& state) {
    for (auto _ : state) {
        gribjump::GribJump gj;
        benchmark::DoNotOptimize(&gj);
    }
}
BENCHMARK(BM_HandleCreation)->Name("cpp/handle_creation");

static void BM_RequestCreation(benchmark::State& state) {
    std::vector<gribjump::Range> ranges = {{0, 10}, {20, 30}};
    std::string request                 = "class=rd,expver=xxxx,stream=oper,date=20230508,time=1200";

    for (auto _ : state) {
        gribjump::ExtractionRequest req(request, ranges);
        benchmark::DoNotOptimize(&req);
    }
}
BENCHMARK(BM_RequestCreation)->Name("cpp/request_creation");

static void BM_RangeCreation(benchmark::State& state) {
    for (auto _ : state) {
        gribjump::Range range = {0, 100};
        benchmark::DoNotOptimize(&range);
    }
}
BENCHMARK(BM_RangeCreation)->Name("cpp/range_creation");

static void BM_RequestManyRanges(benchmark::State& state) {
    // 100 ranges
    std::vector<gribjump::Range> ranges;
    ranges.reserve(100);
    for (int i = 0; i < 100; i++) {
        ranges.push_back({static_cast<size_t>(i * 10), static_cast<size_t>(i * 10 + 5)});
    }
    std::string request = "class=rd,expver=xxxx,stream=oper,date=20230508,time=1200";

    for (auto _ : state) {
        gribjump::ExtractionRequest req(request, ranges);
        benchmark::DoNotOptimize(&req);
    }
}
BENCHMARK(BM_RequestManyRanges)->Name("cpp/request_100_ranges");

BENCHMARK_MAIN();
